### PR TITLE
Server-side throttling PoC

### PR DESCRIPTION
**Pull request by guilload**
### Description
Pushing an early draft to validate the approach.

### How was this PR tested?
- Ingested a bunch of traces with tracegen, the rate limiting logs are gone.

### CODE REVIEW

File: quickwit/quickwit-ingest-api/src/ingest_api_service.rs [hunk 1]
```diff
@@ -90,44 +96,57 @@ impl IngestApiService {
...
+                        .is_ok()
+                    {
+                        if new_capacity > 0 {
+                            self.capacity_notify.notify_one();
+                        }
+                        break;
+                    }
+                }
+            }
+            self.capacity_notify.notified().await;
```
Comment (guilload) [inline quickwit/quickwit-ingest-api/src/ingest_api_service.rs:141]: If truncate is never called, this waits forever :/

Comment (guilload) [reply quickwit/quickwit-ingest-api/src/ingest_api_service.rs:142]: Note to self:
- wrap with timeout and return 429
- count the number of pending clients and return if above > some threshold?

**Review by trinity-1686a — COMMENTED** [review]
LGTM.

I gave it a try, it seemed to work until the rest-client waited too long and decided to timeout itself, but there's a note to self for that

**Comment by guilload** [comment]
Closing in favor of a series of (upcoming) PRs.

**Pull request closed without merging**